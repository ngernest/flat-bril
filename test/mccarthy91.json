{
  "functions": [
    {
      "args": [
        {
          "name": "n",
          "type": "int"
        }
      ],
      "instrs": [
        {
          "dest": "ten",
          "op": "const",
          "type": "int",
          "value": 10
        },
        {
          "dest": "eleven",
          "op": "const",
          "type": "int",
          "value": 11
        },
        {
          "dest": "hundred",
          "op": "const",
          "type": "int",
          "value": 100
        },
        {
          "args": [
            "n",
            "hundred"
          ],
          "dest": "big",
          "op": "gt",
          "type": "bool"
        },
        {
          "args": [
            "big"
          ],
          "labels": [
            "big",
            "small"
          ],
          "op": "br"
        },
        {
          "label": "big"
        },
        {
          "args": [
            "n",
            "ten"
          ],
          "dest": "tmp",
          "op": "sub",
          "type": "int"
        },
        {
          "args": [
            "tmp"
          ],
          "op": "ret"
        },
        {
          "label": "small"
        },
        {
          "args": [
            "n",
            "eleven"
          ],
          "dest": "tmp",
          "op": "add",
          "type": "int"
        },
        {
          "args": [
            "tmp"
          ],
          "dest": "tmp",
          "funcs": [
            "mccarthy91"
          ],
          "op": "call",
          "type": "int"
        },
        {
          "args": [
            "tmp"
          ],
          "dest": "tmp",
          "funcs": [
            "mccarthy91"
          ],
          "op": "call",
          "type": "int"
        },
        {
          "args": [
            "tmp"
          ],
          "op": "ret"
        }
      ],
      "name": "mccarthy91",
      "type": "int"
    },
    {
      "args": [
        {
          "name": "n",
          "type": "int"
        }
      ],
      "instrs": [
        {
          "args": [
            "n"
          ],
          "dest": "result",
          "funcs": [
            "mccarthy91"
          ],
          "op": "call",
          "type": "int"
        },
        {
          "args": [
            "result"
          ],
          "op": "print"
        }
      ],
      "name": "main"
    }
  ]
}
