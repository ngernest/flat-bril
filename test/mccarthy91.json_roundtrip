{
  "functions": [
    {
      "args": [
        {
          "name": "n",
          "type": "int"
        }
      ],
      "instrs": [
        {
          "dest": "ten",
          "op": "const",
          "type": "int",
          "value": 10
        },
        {
          "dest": "eleven",
          "op": "const",
          "type": "int",
          "value": 11
        },
        {
          "dest": "hundred",
          "op": "const",
          "type": "int",
          "value": 100
        },
        {
          "args": [
            "n",
            "hundred"
          ],
          "dest": "big",
          "funcs": [],
          "labels": [],
          "op": "gt",
          "type": "bool"
        },
        {
          "args": [
            "big"
          ],
          "funcs": [],
          "labels": [
            "big",
            "small"
          ],
          "op": "br"
        },
        {
          "label": "big"
        },
        {
          "args": [
            "n",
            "ten"
          ],
          "dest": "tmp",
          "funcs": [],
          "labels": [],
          "op": "sub",
          "type": "int"
        },
        {
          "args": [
            "tmp"
          ],
          "funcs": [],
          "labels": [],
          "op": "ret"
        },
        {
          "label": "small"
        },
        {
          "args": [
            "n",
            "eleven"
          ],
          "dest": "tmp",
          "funcs": [],
          "labels": [],
          "op": "add",
          "type": "int"
        },
        {
          "args": [
            "tmp"
          ],
          "dest": "tmp",
          "funcs": [
            "mccarthy91"
          ],
          "labels": [],
          "op": "call",
          "type": "int"
        },
        {
          "args": [
            "tmp"
          ],
          "dest": "tmp",
          "funcs": [
            "mccarthy91"
          ],
          "labels": [],
          "op": "call",
          "type": "int"
        },
        {
          "args": [
            "tmp"
          ],
          "funcs": [],
          "labels": [],
          "op": "ret"
        }
      ],
      "name": "mccarthy91",
      "type": "int"
    },
    {
      "args": [
        {
          "name": "n",
          "type": "int"
        }
      ],
      "instrs": [
        {
          "args": [
            "n"
          ],
          "dest": "result",
          "funcs": [
            "mccarthy91"
          ],
          "labels": [],
          "op": "call",
          "type": "int"
        },
        {
          "args": [
            "result"
          ],
          "funcs": [],
          "labels": [],
          "op": "print"
        }
      ],
      "name": "main"
    }
  ]
}
