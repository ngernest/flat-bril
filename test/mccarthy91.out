@mccarthy91
	op: const		dest: ten		type: int		value: 10	
	op: const		dest: eleven		type: int		value: 11	
	op: const		dest: hundred		type: int		value: 100	
	op: gt   		dest: big		type: bool		args: ["n", "hundred"]	
	op: br   		args: ["big"]		labels: ["big", "small"]
.big
	op: sub  		dest: tmp		type: int		args: ["n", "ten"]	
	op: ret  		args: ["tmp"]	
.small
	op: add  		dest: tmp		type: int		args: ["n", "eleven"]	
	op: call 		dest: tmp		type: int		args: ["tmp"]		funcs: ["mccarthy91"]
	op: call 		dest: tmp		type: int		args: ["tmp"]		funcs: ["mccarthy91"]
	op: ret  		args: ["tmp"]	
@main
	op: call 		dest: result		type: int		args: ["n"]		funcs: ["mccarthy91"]
	op: print		args: ["result"]	
